### DISCUSSION TITLE
match_input function

### DISCUSSION THREAD

**Discussion by sirdody** [harfbuzz/harfbuzz#3471] [Q&A] [answered]
Hello, 

I am reading harfbuzz source code, particularly match_input function in hb-ot-layout-gsubgpos.cc. I cannot comprehend much from the source code and comment in there. Can you give some example for follow comment:

```
If first component was attached to a previous ligature component,
       * all subsequent components should be attached to the same ligature
       * component, otherwise we shouldn't ligate them...
```
And, is this comment general rule or just a quick hack for some font ligature?

Many thanks,
Toan

**Comment by behdad** [comment, marked as answer]
Hi Toan,

> is this comment general rule or just a quick hack for some font ligature?

It's a general rule.

Let's work an example, where glyphs starting with B are base glyphs, starting with M are mark glyphs, and starting with L are ligature glyphs. If the input sequence is B1,M1,B2,M2,B3,M3, and B1,B2,B3 happen to form a ligature L1, we end up with a glyph sequence L1,M1,M2,M3. Now If there's a rule that forms a ligature L2 out of the sequence M1,M2,M3 however, that sequence should NOT match here, because the marks M1,M2,M3 belong to different ligature-components, to components 1,2,3 respectively. The rule you quoted enforces this.

Does that make sense?
